src/cache-utils.ts

```diff
@@ -66,35 +104,123 @@ export const getPackageManagerInfo = async (packageManager: string) => {
   } else if (packageManager === 'pnpm') {
     return supportedPackageManagers.pnpm;
   } else if (packageManager === 'yarn') {
-    const yarnVersion = await getPackageManagerVersion('yarn', '--version');
-
-    core.debug(`Consumed yarn version is ${yarnVersion}`);
-
-    if (yarnVersion.startsWith('1.')) {
-      return supportedPackageManagers.yarn1;
-    } else {
-      return supportedPackageManagers.yarn2;
-    }
+    return supportedPackageManagers.yarn;
   } else {
     return null;
   }
 };
 
-export const getCacheDirectoryPath = async (
+/**
+ * glob expanding memoized because it involves potentially very deep
+ * traversing through the directories tree
+ */
+export const memoizedCacheDependencies: Record<string, string[]> = {};
+/**
+ * Expands (converts) the string input `cache-dependency-path` to list of directories that
+ * may be project roots
+ * @param cacheDependencyPath - either a single string or multiline string with possible glob patterns
+ *                              expected to be the result of `core.getInput('cache-dependency-path')`
+ * @return list of directories and possible
+ */
+const getProjectDirectoriesFromCacheDependencyPath = async (
+  cacheDependencyPath: string
+): Promise<string[]> => {
+  let cacheDependenciesPaths: string[];
+
+  // memoize unglobbed paths to avoid traversing FS
+  const memoized = memoizedCacheDependencies[cacheDependencyPath];
+  if (memoized) {
+    cacheDependenciesPaths = memoized;
+  } else {
+    const globber = await glob.create(cacheDependencyPath);
+    cacheDependenciesPaths = (await globber.glob()) || [''];
+    memoizedCacheDependencies[cacheDependencyPath] = cacheDependenciesPaths;
+  }
+
+  const existingDirectories: string[] = cacheDependenciesPaths
+    .map(path.dirname)
+    .filter(path => path != null)
+    .filter(unique())
+    .filter(fs.existsSync)
+    .filter(directory => fs.lstatSync(directory).isDirectory());
```

Comment:
Lets combine this filters.

Reply:
@nikolai-laevskii  your opnion?

Reply:
@dmitry-shibanov I previously asked to separate these filters for better readability. Is there a reason why we would want them to be combined?